packages/runtime/src/errors/errors-list.ts

```diff
@@ -207,9 +207,15 @@ by an index that does not exist.`
   },
   UNKOWN_OPCODE: {
     number: 31,
-    message: "Error encountered while parsing teal file unkown opcode %opcode% , Line : %line% ",
+    message: "Error encountered while parsing teal file: unkown opcode \"%opcode%\" for teal version #%version%, Line : %line% ",
     title: PARSE_ERROR,
     description: `Unkown Opcode encountered`
+  },
+  INVALID_LOGICSIG_MAX_COST: {
```

Comment:
Is this only for  logic signature or stateful too ? or are they same in this case ?

Reply:
https://developer.algorand.org/docs/reference/teal/specification/
just saw this section In `Execution Mode`, it is mentioned we have different COST for both MaxAppProgramCost and LogicSigMaxCost. can you please confirm if their values are same ?

Reply:
also there is Max program length (consensus parameters LogicSigMaxSize, MaxApprovalProgramLen and MaxClearStateProgramLen) different for different programs

Reply:
I was checking the statement above the `Execution Mode` and thought it was generic.
![image](https://user-images.githubusercontent.com/33264364/105207441-8bcd9100-5b6d-11eb-86d4-fc14498ab03e.png)

I think you are right, starting from version 2, stateless and stateful teal have different max cost. But i had question regarding this, currently can we know in parser whether TEAL is stateless or stateful?

Reply:
I don't think that's a good idea, because clear program doesn't have ant stateful opcode but it is stateful.(possible for approval too).
One other way would be:
I think we can run the parser to store interpreter.gas(as we do now). the only way to know if current TEAL is stateless, approval or clear is via execParams, We can add a check in interpreter which will throw error according to TransactionType or SignType

Reply:
after parsing is done, we will have interpreter.gas, which we can use in interpreter to check for max gas, and throw error accordingly

Reply:
first we can check for `appID` if it is passed, then program is approval or  clear
if not we can check for SignType
If it is logicSig then we proceed with LogicSigMaxCost.

Reply:
I don't think checking by `appId` is a good approach (it's not intuitive)
Extending you previous comment, we can do:
if SignType is LogicSignature, then it's stateless TEAL - `LogicSigMaxCost`, `LogicSigMaxSize`
If not, we can check if TransactionType is `ClearSSC` (For clear program) - `MaxAppProgramCost`, `MaxApprovalProgramLen`
if TransactionType is `NoOp | CloseSSC | DeleteSSC` (For approval program) - `MaxAppProgramCost`, `MaxClearStateProgramLen`

Reply:
Also, i think we can support this for a single tx for now. For group transaction we currently support only passing a single TEAL file (for stateless + stateful), whereas we should be able to pass teal code with each stateless tx (as they only approve/reject a txn).